### PR TITLE
initial code

### PR DESCRIPTION
**Pull request by go-oleg**
Some notes
- Borrowed some code from https://github.com/atom/markdown-preview because we want a similar behavior w.r.t. to opening a new tab in the right pane when running a juttle or switching to an existing tab if the juttle has already been run.
- I couldn't get the styles from juttle-client-library to load properly using an import so copied them from `node_modules` into `styles` and checked it in.

@mnibecker

### CODE REVIEW

File: lib/juttle-viewer-view.coffee [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+module.exports =
+class JuttleViewerView extends ScrollView
+  @content: ->
+    @div class: 'juttle-viewer native-key-bindings', tabindex: -1
+
+  constructor: (@editorId) ->
+    super
+    juttleClient = new JuttleClient(atom.config.get('juttle-viewer.juttleServiceHost'))
+    @juttleClientView = new juttleClient.View(@element);
```
Comment (mattnibs) [inline lib/juttle-viewer-view.coffee]: You might want to do this with `juttle-viewer@0.7.0-rc.5` as the invocation changes.

File: lib/juttle-viewer-view.coffee [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  getTitle: ->
+    path = @editorForId(@editorId)?.getPath()
+    if path then path.substring(path.lastIndexOf('/') + 1) else '[unknown]'
+
+  getURI: ->
+    return "juttle-viewer://editor/#{@editorId}"
+
+  run: ->
+    text = @editorForId(@editorId).getText()
+    @juttleClientView.run({ program: text }).catch (err) =>
```
Comment (mattnibs) [inline lib/juttle-viewer-view.coffee]: feels like if we're not supporting inputs we might want to at least warn them when running programs will inputs. You could do this here by adding an `api.describe` and adding a warning if `inputs.length > 0`

File: lib/juttle-viewer-view.coffee [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    for editor in atom.workspace.getTextEditors()
+      return editor if editor.id?.toString() is editorId.toString()
+    null
+
+  getTitle: ->
+    path = @editorForId(@editorId)?.getPath()
+    if path then path.substring(path.lastIndexOf('/') + 1) else '[unknown]'
+
+  getURI: ->
+    return "juttle-viewer://editor/#{@editorId}"
```
Comment (mattnibs) [inline lib/juttle-viewer-view.coffee:36]: how is this used?

File: lib/juttle-viewer-view.coffee [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  constructor: (@editorId) ->
+    super
+    juttleClient = new JuttleClient(atom.config.get('juttle-viewer.juttleServiceHost'))
+    @juttleClientView = new juttleClient.View(@element);
+
+    @juttleClientView.on 'error', (err) =>
+      @_showError(err)
+
+    @juttleClientView.on 'warning', (warn) =>
+      @_showWarning(warn)
```
Comment (mattnibs) [inline lib/juttle-viewer-view.coffee]: There doesn't seem to be a deconstructor for this. If you were to open up multiple juttles the events would never get unsubscribe these events and you would get a memory leak, right?

Comment (mattnibs) [inline lib/juttle-viewer-view.coffee]: actually opening mulitple juttles, but this should have a deconstructor when they close `juttle-viewer`

File: lib/juttle-viewer.coffee [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+  uriForEditor: (editor) ->
+    "juttle-viewer://editor/#{editor.id}"
+
+  createJuttleViewerView: (state) ->
+    JuttleViewerView ?= require './juttle-viewer-view'
+    new JuttleViewerView(state)
+
+  deactivate: ->
+    @subscriptions.dispose()
```
Comment (mattnibs) [inline lib/juttle-viewer.coffee:33]: unsubscribe view events here

**Comment by mattnibs** [comment]
Before merge:
I would use 0.7.0 jcl instead, deconstruct your view object on close, and also check for inputs in a program, and also fix the annoying dropdown css things

File: lib/juttle-viewer-view.coffee (hunk 3 above)
Comment (go-oleg) [inline lib/juttle-viewer-view.coffee:36]: This is what is used by `atom.workspace.addOpener` to see if the editor you are running the juttle from already has an output tab opened for it.

File: lib/juttle-viewer.coffee (hunk 5 above)
Comment (go-oleg) [inline lib/juttle-viewer.coffee:33]: Fixed to unsubscribe in juttle-viewer-view.destroy.

File: lib/juttle-viewer-view.coffee (hunk 4 above)
Comment (go-oleg) [inline lib/juttle-viewer-view.coffee]: Done in `destroy()`

File: lib/juttle-viewer-view.coffee (hunk 2 above)
Comment (go-oleg) [inline lib/juttle-viewer-view.coffee]: Good call. Made it so it shows the warning once.

**Comment by go-oleg** [comment]
@mnibecker addressed your feedback, take a look? (will squash afterwards)

**Comment by mattnibs** [comment]
+1

**Pull request merged**